### PR TITLE
Key improvements

### LINKED ISSUE #3107: ValueKey should use runtimeType to determine equality
Not `is ValueKey`, which will collide with subclasses

### PR DESCRIPTION
**Pull request by Hixie**
ValueKey and ObjectKey shouldn't be == with subclasses.
Fixes https://github.com/flutter/flutter/issues/3107

Clean up toString for the keys a bit.

Add a test for keys.

### CODE REVIEW

File: packages/flutter/lib/src/widgets/framework.dart [hunk 1]
```diff
@@ -61,19 +67,28 @@ class ValueKey<T> extends LocalKey {
...
-  int get hashCode => value.hashCode;
+  int get hashCode => hashValues(runtimeType, value);
 
   @override
-  String toString() => '[\'$value\']';
+  String toString() {
+    final String valueString = T == String ? '<\'$value\'>' : '<$value>';
+    // The crazy on the next line is a workaround for
+    // https://github.com/dart-lang/sdk/issues/28548
+    if (runtimeType == new _TypeLiteral<ValueKey<T>>().type)
```
Comment (abarth) [inline packages/flutter/lib/src/widgets/framework.dart]: const ?

**Review by abarth — APPROVED** [review]
LGTM

File: packages/flutter/lib/src/widgets/framework.dart (hunk 1 above)
Comment (Hixie) [inline packages/flutter/lib/src/widgets/framework.dart]: You can't use const, because the T isn't const.

Comment (abarth) [inline packages/flutter/lib/src/widgets/framework.dart]: Makes sense.

**Pull request merged**